### PR TITLE
[tensor] refactor param op hook

### PR DESCRIPTION
**Pull request by ver217**
`torch.autograd.Function` ensures the device of original inputs and grads are the same. Since zero hook may change the device of input in pre fwd, we execute pre fwd out of `autograd.Function`.

### CODE REVIEW

**Comment by FrankLeeeee** [comment]
Can you add some documentation for better readability and maintainability.

**Comment by ver217** [comment]
> Can you add some documentation for better readability and maintainability?

DONE

File: colossalai/tensor/param_op_hook.py [hunk 1]
```diff
@@ -23,25 +28,78 @@ def post_backward(self, params: List[torch.Tensor]) -> None:
...
 
+    @staticmethod
+    @contextmanager
+    def use_hooks(*hooks: ParamOpHook):
+        """Change the param op hooks you use. Nested calling is allowed.
+
+        Example::
+            >>> with ParamOpHookManager.use_hooks(*hooks):
+            >>>     do_something()
+            >>>     with ParamOpHookManager.use_hooks():
```
Comment (feifeibear) [inline colossalai/tensor/param_op_hook.py:45]: why nest ParamOpHookManager context?

Comment (ver217) [reply colossalai/tensor/param_op_hook.py:45]: In case some users want to nest. In most cases, we don't need to nest.

Comment (feifeibear) [reply colossalai/tensor/param_op_hook.py:45]: I think users should not know what is OpHook. It is not a common-sense concept. Why do they need to have a manager?

Comment (ver217) [reply colossalai/tensor/param_op_hook.py:45]: They can extend profiler by this manager. E.g. before op sample sth, after op sample sth. Indeed, it's for advanced developers.

**Review by feifeibear — APPROVED** [review]

**Pull request merged**